sklearn/ensemble/partial_dependence.py

```diff
@@ -306,8 +306,8 @@ def convert_feature(fx):
                 l.append(feature_names[i])
             names.append(l)
     except IndexError:
-        raise ValueError('features[i] must be in [0, n_features) '
-                         'but was %d' % i)
+        raise ValueError('features[i] must be in [0, len(feature_names)={0} ) '
+                         'but was {1}'.format(len(feature_names), i))
```

Comment:
Unless I'm missing something, after the `features` is converted by the prior code, the looping below causes  `i` to be a value from the original `features` array that was passed in (if `features` containing integer values).

```
for fxs in features:
            l = []
            # explicit loop so "i" is bound for exception below
            for i in fxs:
                l.append(feature_names[i])
            names.append(l)
    except IndexError:
    ...
```

Part of the issue may come from reassigning the `features` parameter that was passed in to `tmp_features`. After that, any reference to `features[i]` in the code takes a different meaning than what a user thinks of when reading `features[i]`.